### PR TITLE
refactor: Create EnergyLevels and EnergyHandler classes for dealing with Ship energies

### PR DESCRIPTION
**Pull request by Amazinite**
**Refactor:** This PR implements the refactor detailed in issue #6576.

## Refactor Details
This PR refactors Ship as described in #6576 and partially in #6329 by adding two new classes:
* EnergyLevels: this class is nothing but a collection of doubles that represent various "energy levels" that a ship has. This includes hull, shields, heat, fuel, energy, and the DoT variables. Ship now contains an EnergyLevels object instead of a field for each of the levels. The name for this class comes from the following comment:
https://github.com/endless-sky/endless-sky/blob/3882d31009f52bce40e238db34a4a5970565f796/source/Ship.h#L512-L513
* EnergyHandler: this class is responsible for some (but not all) of the handling of a ship's EnergyLevels. EnergyHandler is currently responsible for repairs, taking damage, applying and resisting DoT status effects, keeping track of movement costs (i.e. thrust, reverse thrust, turning, afterburners), cloaking, evaluating the costs of firing weapons, and keeping track of capacities and minimum hull values.
  * The EnergyHandler does this by containing a set of EnergyLevels objects that each record the cost of these various actions. These EnergyLevels are updated whenever Ship::AddOutfit is called, meaning they are always up-to-date with the Ship's actual attributes. By storing costs in these EnergyLevels instead of recalling attributes.Get() every frame, EnergyHandler allows for cutting down on the number of string comparisons made each frame, as mentioned in #6329. 
  * For weapons, the EnergyLevels cost of firing the weapon is created on the fly. This means that there's no performance improvement from this change, but there is an organizational improvement by moving quite a few lines of code from Ship.cpp to EnergyHandler.cpp.

## Testing Done
Flew around with these changes. Observed how thrusting, reverse thrusting, turning, and afterburning all still work. Ships still take damage, repair, and bleed DoT.

## Performance Impact
This change cuts down on string comparisons by duplicating some attribute values from the attributes Outfit in Ship to various EnergyLevels in EnergyHandler, the tradeoff being memory usage from the duplicate doubles.

EnergyHandler updates its EnergyLevels whenever Ship::AddOutfit is called, meaning that adding an outfit now does many more string comparison than before, but I believe that AddOutfit is only ever called when in the outfitter or while boarding, which aren't performance intensive areas to begin with, so we should be good. (And these same string comparison are currently being done every frame for every ship anyway, so moving them to happen once whenever AddOutfit is called will be a performance improvement overall.)

### CODE REVIEW

**Comment by Amazinite** [comment]
Pretty sure this refactor also fixes a couple bugs/missing features:
* Weapons use to only check firing ion, disruption, and slowing. They now check all DoT costs.
* Weapons use to apply firing shields in full when fired. This caused ships to gain negative shields, which shouldn't be allowed to happen. Weapons now only apply the firing shield cost up to the amount of remaining shields.
* When thrusting or turning and applying fractional usage, if multiple costs were lacking but a later cost was closer to full than an earlier evaluated cost, the scale would be changed to match the later cost, therefore going over the budget of the earlier cost. Evaluating fractional thrust/turn now takes the rolling scale into account at each step, meaning that no costs should ever go over budget.
* Engines now check thrusting/turning DoT as a cost, whereas before they didn't. This matches their behavior to weapons, and prevents DoT counters from going negative. Question: should we allow engines to fire when they have a negative firing DoT, even if there's no DoT to remove? We'd them clamp any firing DoT to the remaining DoT counter, just as described above with weapon firing shields and the ship's remaining shields.

**Comment by waterhouse** [comment]
Profiling this code, using the Mega Bucks pilotfile I put in https://github.com/endless-sky/endless-sky/issues/6329 , this looks to be a significant improvement.  In AI::AutoFire, now the "heaviest stack trace" is Mask::Collide (which looks like "real, necessary" computation), instead of the Dictionary::Get calls in Ship::IsDisabled.  Overall time spent in Dictionary::Get has dropped from about 30% to about 10%.

A breakdown of most of the remaining Dictionary::Get time, from a particular sample:
- 2.1% in IsStranded (1.2% JumpFuel, 0.7% JumpsRemaining)
- 1.3% in RequiredCrew
- 2% in Ship::Move (1.4% in DoGeneration)
- 1.3% in EnergyHandler::FiringCost

**Comment by Amazinite** [comment]
A number of the remaining Gets in DoGeneration could probably be handled by EnergyHandler (e.g. energy generation, heat generation). The number of attributes that EnergyHandler wouldn't deal with are probably few and far enough between (e.g. required crew) that they could just get their own fields that are updated separately. The only doozy would be jump fuel and range, since those are dependent on individual outfits; perhaps ships could have a map of ranges and fuel consumptions for that range for each possible range given the ship's outfits. And FiringCost is going to need to stay how it is unless we store the FiringCost of each weapon when it's installed instead of generating it on the fly. Could probably have Armament keep track of FiringCosts.

**Comment by warp-core** [comment]
Is this just waiting on a review from Tehhowch?

**Review by warp-core — APPROVED** [review]
Oh, I have to click this button for these comments to be visible to other people. Oops.
I'm not suggesting you make these changes here, this still seems a lot better than what we have now and I'm not entirely sure that my concept would work or be better.
Not sure my approval means much if you look at my PRs, but here you go.

File: source/EnergyHandler.cpp
Comment (tibetiroka) [inline source/EnergyHandler.cpp:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```
The standard copyright header was changed recently.

File: source/EnergyHandler.h
Comment (tibetiroka) [inline source/EnergyHandler.h:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

File: source/EnergyLevels.h
Comment (tibetiroka) [inline source/EnergyLevels.h:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

**Review by tibetiroka — CHANGES_REQUESTED** [review]

**Comment by Amazinite** [comment]
This isn't a super important refactor so I'm gonna close this and revisit the idea later while there's less stuff on my plate.

**Pull request closed without merging**
